File: app-paper/src/main/java/com/paper/exp/simulation/CollisionSystemView.java
package com.paper.exp.simulation;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import com.paper.R;
import com.paper.exp.simulation.CollisionSystemContract.SimulationListener;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import io.reactivex.Observable;

public class CollisionSystemView
    extends View
    implements CollisionSystemContract.View {

    // Rendering.
    private final Paint mParticlePaint = new Paint();
    private final Paint mTextPaint = new Paint();
    private final Paint mBoundPaint = new Paint();
    private final RectF mBoundRect = new RectF();
    private final RectF mOval = new RectF();
    private float mTextSize = 0f;

    private SimulationListener mListener;

    public CollisionSystemView(Context context) {
        super(context);
    }

    public CollisionSystemView(Context context,
                               @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CollisionSystemView(Context context,
                               @Nullable AttributeSet attrs,
                               int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        mParticlePaint.setStyle(Paint.Style.FILL);
        mParticlePaint.setColor(Color.BLACK);

        mTextSize = context.getResources().getDimension(R.dimen.debug_text_size_1);

        mTextPaint.setColor(Color.GREEN);
        mTextPaint.setTextSize(mTextSize);
        mTextPaint.setTextAlign(Paint.Align.LEFT);

        mBoundPaint.setColor(Color.LTGRAY);
        mBoundPaint.setStyle(Paint.Style.FILL);
    }

    @Override
    public void schedulePeriodicRendering(@NotNull SimulationListener listener) {
        mListener = listener;
        postInvalidate();
    }

    @Override
    public void unScheduleAll() {
        mListener = null;
    }

    @NotNull
    @Override
    public Observable<Object> onClickBack() {
        return Observable.just((Object) 0);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Protected / Private Methods ////////////////////////////////////////////

    @Override
    protected void onMeasure(int widthSpec,
                             int heightSpec) {
        final int width = MeasureSpec.getSize(widthSpec);
        final int height = width;

        mBoundRect.set(0, 0, width, height);

        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        // Boundary.
        canvas.drawRect(mBoundRect, mBoundPaint);

        if (mListener != null) {
            mListener.onUpdateSimulation(canvas);
            postInvalidate();
        }
    }

    @Override
    public void showToast(@NotNull String text) {
        // DUMMY.
    }

    @Override
    public void drawDebugText(@NotNull Canvas canvas,
                              @NotNull String text) {
        float x = mTextSize;
        float y = mTextSize;
        for (String line : text.split("\n")) {
            canvas.drawText(line, x, y, mTextPaint);
            y += mTextSize;
        }
    }

    @Override
    public void drawParticles(@NotNull Canvas canvas,
                              @NotNull List<Particle> particles) {
        final int canvasWidth = getWidth();
        final int canvasHeight = getHeight();

        for (int i = 0; i < particles.size(); ++i) {
            final Particle particle = particles.get(i);
            final double x = particle.getCenterX();
            final double y = particle.getCenterY();
            final double r = particle.getRadius();

            // Paint first one in red and the rest in black.
            if (i == 0) {
                mParticlePaint.setColor(Color.RED);
            } else {
                mParticlePaint.setColor(Color.BLACK);
            }

            mOval.set((float) ((x - r) * canvasWidth),
                      (float) ((y - r) * canvasHeight),
                      (float) ((x + r) * canvasWidth),
                      (float) ((y + r) * canvasHeight));

            canvas.drawOval(mOval, mParticlePaint);
        }
    }
}

File: app-presentation/src/main/java/com/paper/view/collisionSimulation/CollisionSystemView.kt
// Copyright Apr 2017-present Paper
//
// Author: boyw165@gmail.com
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

package com.paper.view.collisionSimulation

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View

import com.paper.R

import io.reactivex.Observable

class CollisionSystemView : View {

    // Rendering.
    private val mParticlePaint = Paint()
    private val mTextPaint = Paint()
    private val mBoundPaint = Paint()
    private val mBoundRect = RectF()
    private val mOval = RectF()
    private var mTextSize = 0f
    private var mTextPadding = 0f

    private var mListener: SimulationListener? = null

    val canvasWidth: Int
        get() = width

    val canvasHeight: Int
        get() = height

    constructor(context: Context) : super(context) {}

    @JvmOverloads constructor(context: Context,
                              attrs: AttributeSet?,
                              defStyleAttr: Int = 0) : super(context, attrs, defStyleAttr) {

        mParticlePaint.style = Paint.Style.FILL
        mParticlePaint.color = Color.BLACK

        mTextSize = context.resources.getDimension(R.dimen.debug_text_size_1)
        mTextPadding = context.resources.getDimension(R.dimen.debug_padding)

        mTextPaint.color = Color.parseColor("#006400")
        mTextPaint.textSize = mTextSize
        mTextPaint.textAlign = Paint.Align.LEFT

        mBoundPaint.color = Color.LTGRAY
        mBoundPaint.style = Paint.Style.FILL
    }

    fun schedulePeriodicRendering(listener: SimulationListener) {
        mListener = listener
        postInvalidate()
    }

    fun unScheduleAll() {
        mListener = null
    }

    fun onClickBack(): Observable<Any> {
        return Observable.just(0 as Any)
    }

    ///////////////////////////////////////////////////////////////////////////
    // Protected / Private Methods ////////////////////////////////////////////

    override fun onMeasure(widthSpec: Int,
                           heightSpec: Int) {
        val width = View.MeasureSpec.getSize(widthSpec)
        val height = width

        mBoundRect.set(0f, 0f, width.toFloat(), height.toFloat())

        setMeasuredDimension(width, height)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // Boundary.
        canvas.drawRect(mBoundRect, mBoundPaint)

        if (mListener != null) {
            mListener!!.onUpdateSimulation(canvas)
            postInvalidate()
        }
    }

    fun showToast(text: String) {
        // DUMMY.
    }

    fun drawDebugText(canvas: Canvas,
                      text: String) {
        val x = mTextPadding
        var y = mTextPadding + mTextSize / 2f
        for (line in text.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) {
            canvas.drawText(line, x, y, mTextPaint)
            y += mTextSize
        }
    }

    fun drawParticles(canvas: Canvas,
                      particles: List<Particle>) {
        val canvasWidth = canvasWidth
        val canvasHeight = canvasHeight

        for (i in particles.indices) {
            val particle = particles[i]
            val x = particle.centerX
            val y = particle.centerY
            val r = particle.radius

            // Paint first one in red and the rest in black.
            if (i == 0) {
                mParticlePaint.color = Color.RED
            } else {
                mParticlePaint.color = Color.BLACK
            }

            mOval.set(((x - r) * canvasWidth).toFloat(),
                      ((y - r) * canvasHeight).toFloat(),
                      ((x + r) * canvasWidth).toFloat(),
                      ((y + r) * canvasHeight).toFloat())

            canvas.drawOval(mOval, mParticlePaint)
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // Clazz //////////////////////////////////////////////////////////////////

    interface SimulationListener {

        fun onUpdateSimulation(canvas: Canvas)
    }
}
